lib/src/impl_stub/impl_stub.dart

```diff
@@ -0,0 +1,11 @@
+library impl_stub;
+
+import 'dart:typed_data';
+import 'dart:async';
+
+import '../impl_stub.dart'
+    if (dart.library.ffi) '../impl_ffi/impl_ffi.dart'
+    if (dart.library.js) '../impl_js/impl_js.dart' as impl;
```

Comment:
okay, I forgot we had `lib/src/impl_stub.dart` already. So `lib/src/impl_stub/...` is perhaps a bad name.

I'll try to outline what we have:

 * `lib/webcrypto.dart` exports public classes from `lib/src/webcrypto/webcrypto.dart`, all classes exported should be `final`.
 * `lib/src/webcrypto/webcrypto.dart` contains the implementation of the public classes (these will be `final`).
 * `lib/src/impl_stub.dart`, contains a "stub" platform implementation.
    This implements everything that a platform implementation must provide, it just does it in the trivial way where every method throws `UnimplementedError`.
    This platform implementation is kind of useless, but it's technically a correct implementation :rofl: 
 * `lib/src/impl_js/impl_js.dart`, contains a platform implementation using js.
    This library has the same interface (top-level methods) as `lib/src/impl_stub.dart`.
 * `lib/src/impl_js/impl_ffi.dart`, contains a platform implementation using ffi.
    This library has the same interface (top-level methods) as `lib/src/impl_stub.dart`.


So `lib/src/webcrypto/webcrypto.dart` needs to import either:
 * `lib/src/impl_stub.dart`,
 * `lib/src/impl_js/impl_js.dart`, or.
 * `lib/src/impl_js/impl_ffi.dart`.

They all have the same top-level methods and interface, so we can use configurable imports.

------------------------

What we're trying to do in `lib/src/impl_stub/impl_stub.dart` is to create a library the specifies the interfaces that platform libraries should implement. That is interfaces that should be implemented by:
 * `lib/src/impl_stub.dart`,
 * `lib/src/impl_js/impl_js.dart`, and,
 * `lib/src/impl_js/impl_ffi.dart`.

Perhaps calling it `lib/src/impl_stub/impl_stub.dart` is a mistake. Maybe, it's better to call it `impl_interface` ?

After all we want an import-graph where:
 * `lib/webcrypto.dart` exports:
 * `lib/src/webcrypto/webcrypto.dart` which imports, one of:
   * `lib/src/impl_stub.dart`,
   * `lib/src/impl_js/impl_js.dart`, or
   * `lib/src/impl_js/impl_ffi.dart`.
   Each of which imports:
 * `lib/src/impl_interface/impl_interface.dart`.


-------

Maybe, (this is just an idea) we should make:
```dart
# lib/src/impl_interface/impl_interface.dart

/// A platform implementation must provide a public top-level constant implementation
/// of [WebCryptoImpl] called `webCryptImpl`.
abstract interface class WebCryptoImpl {
  StaticAesCbcSecretKeyImpl get aesCbcSecretKey;
}

# lib/src/impl_interface/impl_interface.aescbc.dart

abstract interface class StaticAesCbcSecretKeyImpl {
  Future<AesCbcSecretKeyImpl> importRawKey(List<int> keyData);
  Future<AesCbcSecretKeyImpl> importJsonWebKey(Map<String, dynamic> jwk);
  Future<AesCbcSecretKeyImpl> generateKey(int length);
}

abstract interface class AesCbcSecretKeyImpl {
  Future<Uint8List> encryptBytes(List<int> data, List<int> iv);
  Future<Uint8List> decryptBytes(List<int> data, List<int> iv);
  Stream<Uint8List> encryptStream(Stream<List<int>> data, List<int> iv);
  Stream<Uint8List> decryptStream(Stream<List<int>> data, List<int> iv);
  Future<Uint8List> exportRawKey();
  Future<Map<String, dynamic>> exportJsonWebKey();
}

# lib/src/impl_stub/impl_stub.dart  // this is now replacing lib/src/impl_stub.dart

const webCryptImpl = _WebCryptoImpl();

final class _WebCryptoImpl implememts WebCryptoImpl {
  const _WebCryptoImpl();

  final aesCbcSecretKey = _StaticAesCbcSecretKeyImpl();
}

# lib/src/impl_stub/impl_stub.aescbc.dart

final class _StaticAesCbcSecretKeyImpl implements StaticAesCbcSecretKeyImpl {
  const _StaticAesCbcSecretKeyImpl();

  @override
  Future<AesCbcSecretKeyImpl> importRawKey(List<int> keyData) =>
    throw UnimplementedError('not implemented');

  @override
  Future<AesCbcSecretKeyImpl> importJsonWebKey(Map<String, dynamic> jwk) =>
    throw UnimplementedError('not implemented');

  @override
  Future<AesCbcSecretKeyImpl> generateKey(int length) =>
    throw UnimplementedError('not implemented');
}

# lib/src/impl_js/impl_js.dart

const webCryptImpl = _WebCryptoImpl();

final class _WebCryptoImpl implememts WebCryptoImpl {
  const _WebCryptoImpl();

  final aesCbcSecretKey = _StaticAesCbcSecretKeyImpl();
}

# lib/src/impl_js/impl_js.aescbc.dart

final class _StaticAesCbcSecretKeyImpl implements StaticAesCbcSecretKeyImpl {
  const _StaticAesCbcSecretKeyImpl();

  @override
  Future<AesCbcSecretKeyImpl> importRawKey(List<int> keyData) async {
     ...
     
     return _AesCbcSecretKeyImpl(...)
  }

  @override
  Future<AesCbcSecretKeyImpl> importJsonWebKey(Map<String, dynamic> jwk) => ...

  @override
  Future<AesCbcSecretKeyImpl> generateKey(int length) => ...
}

final class _AesCbcSecretKeyImpl implements AesCbcSecretKeyImpl {
  ...
}

# lib/src/impl_ffi/impl_ffi.dart         // similar to impl_ffi.dart
# lib/src/impl_ffi/impl_ffi.aescbc.dart  // similar to impl_ffi.aescbc.dart

# lib/src/webcrypto/webcrypto.dart

import '../impl_interface/impl_interface.dart'
import '../impl_stub/impl_stub.dart'
    if (dart.library.ffi) '../impl_ffi/impl_ffi.dart'
    if (dart.library.js) '../impl_js/impl_js.dart';

# lib/src/webcrypto/webcrypto.aescbc.dart

final class AesCbcSecretKey {
  final AesCbcSecretKeyImpl _impl;

  AesCbcSecretKey._(this._impl); 
  
  static Future<AesCbcSecretKey> importRawKey(List<int> keyData) async {
    final impl = await webCryptImpl.aesCbcSecretKey.importRawKey(keyData);
    return AesCbcSecretKey._(impl);
  }

  ...

  Future<Uint8List> encryptBytes(List<int> data, List<int> iv) =>
      _impl.encryptBytes(data, iv);
  ...
}
```

Maybe, this is crazy overkill... it certainly could be a bit slower due to all the indirection, but since all the indirection is through constants, it might be not affect performance.

If we did this, then the top-level constant `webCryptImpl` is the only thing exported from:
 * `lib/src/impl_stub.dart`,
 * `lib/src/impl_js/impl_js.dart`, and
 * `lib/src/impl_js/impl_ffi.dart`.

And then the type checker will do a pretty decent job at making sure all implementations actually provide a `webCryptImpl` that correctly implements `WebCryptoImpl`.

So we don't end up relying so much on all these implementations providing the same top-level methods, because they just provide one top-level constant, that implements all the methods.

If we were really paranoid we could even do something like:
``` 
# lib/src/webcrypto/webcrypto.aescbc.dart

final class AesCbcSecretKey {
  final AesCbcSecretKeyImpl _impl;

  AesCbcSecretKey._(this._impl); 
  
  static Future<AesCbcSecretKey> importRawKey(List<int> keyData) async {
    assert(webCryptImpl is WebCryptoImpl); // Compiler should hopefully (who knows) be smart enough to compile this away if it's true, since webCryptImpl is a constant
    final impl = await webCryptImpl.aesCbcSecretKey.importRawKey(keyData);
    return AesCbcSecretKey._(impl);
  }

  ...

  Future<Uint8List> encryptBytes(List<int> data, List<int> iv) =>
      _impl.encryptBytes(data, iv);
  ...
}
```

Reply:
Updated to the new architecture. Removed relevant methods from `src/impl_stub.dart`. Will phase out this file sequentially as all classes are refactored 🚀